pkg/controllerhelpers/scylladbmanager.go

```diff
@@ -34,3 +39,139 @@ func GetScyllaDBManagerClient(_ context.Context, _ *scyllav1alpha1.ScyllaDBManag
 func IsManagedByGlobalScyllaDBManagerInstance(smcr *scyllav1alpha1.ScyllaDBManagerClusterRegistration) bool {
 	return naming.GlobalScyllaDBManagerClusterRegistrationSelector().Matches(labels.Set(smcr.GetLabels()))
 }
+
+const (
+	authTokenSize = 128
+)
+
+type ScyllaDBManagerAgentCustomConfigError struct {
+	error
+}
+
+func (e ScyllaDBManagerAgentCustomConfigError) Unwrap() error {
+	return e.error
+}
+
+func NewScyllaDBManagerAgentCustomConfigError(err error) error {
+	return ScyllaDBManagerAgentCustomConfigError{err}
+}
+
+var _ error = (*ScyllaDBManagerAgentCustomConfigError)(nil)
+
+// GetScyllaDBManagerAgentAuthTokenConfigOptions defines options for selecting ScyllaDB Manager agent auth token config.
+type GetScyllaDBManagerAgentAuthTokenConfigOptions struct {
+	GetOptionalCustomAgentConfigSecret func() ([]metav1.Condition, *corev1.Secret, error)
+	GetOptionalExistingAuthTokenSecret func() ([]metav1.Condition, *corev1.Secret, error)
+	// ContinueOnCustomAgentConfigError allows the controller to continue on an error coming from GetOptionalCustomAgentConfigSecret func or extracting the auth token from the custom agent config secret.
+	// This is provided for backwards compatibility, so that a misconfigured or missing custom agent config does not block the controller from creating the auth token secret.
+	// Its use in new applications is discouraged, as it leads to confusing behavior.
+	ContinueOnCustomAgentConfigError bool
+}
+
+func GetScyllaDBManagerAgentAuthTokenConfig(
+	options GetScyllaDBManagerAgentAuthTokenConfigOptions,
+) ([]metav1.Condition, []byte, error) {
+	return getScyllaDBManagerAgentAuthTokenConfig(
+		func() string {
+			return apimachineryutilrand.String(authTokenSize)
+		},
+		options,
+	)
+}
+
+func getScyllaDBManagerAgentAuthTokenConfig(
+	generateAuthToken func() string,
+	options GetScyllaDBManagerAgentAuthTokenConfigOptions,
+) ([]metav1.Condition, []byte, error) {
+	var customConfigError error
+
+	progressingConditions, authToken, err := getScyllaDBManagerAgentAuthToken(
+		generateAuthToken,
+		options,
+	)
+	if err != nil {
+		err = fmt.Errorf("can't get ScyllaDB Manager agent auth token: %w", err)
+		if !options.ContinueOnCustomAgentConfigError || !errors.As(err, &ScyllaDBManagerAgentCustomConfigError{}) {
+			return progressingConditions, nil, err
+		}
+
+		customConfigError = err
+	}
+	if len(progressingConditions) > 0 {
+		return progressingConditions, nil, customConfigError
+	}
+
+	authTokenConfig, err := helpers.GetAgentAuthTokenConfig(authToken)
+	if err != nil {
+		return nil, nil, fmt.Errorf("can't get ScyllaDB Manager agent auth token config: %w", err)
```

Comment:
Shouldn't we return an aggregated error here? If `customConfigError != nil`, it won't be surfaced.

Reply:
done